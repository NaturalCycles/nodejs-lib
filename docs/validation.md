# Validation

Based on popular [Joi](https://github.com/sideway/joi) validation library.

Assumes that you don't need to have `joi` as your dependency, but just use convenience methods from
`nodejs-lib` (and use exported `Joi`, which is extended with all the extensions and defaults
documented below). Do not import `joi` directly, otherwise you lose these defaults and extensions.

Default options:

- `presense: required` - all declared fields are required by default
- `convert: true`
- `allowUnknown: true`
- `stripUnknown:`
  - `objects: true`
  - `arrays: false`

Extensions:

- Strings are trimmed by default
- Treats empty strings `'''` as empty/undefined
- `dateString` stringSchema extension to validate ISO8601 dates
- `dividable` numberSchema extension to validate divide-ability

Does some processing of the Error thrown:

- Includes `objectName` (if provided), otherwise tries to detect "classname" (doing
  `value?.constructor?.name`).
- Includes `objectId` if it can be detected (`id` field of an object passed). Useful e.g in `db-lib`
  CommonDao methods.
- Truncates error message to 1000 characters.
- Truncates `err.details` to max 5 items

## validate

- Validates with Joi.
- Throws JoiValidationError if invalid.
- Returns _converted_ value.
- If `schema` is undefined - returns value as is.

```ts
validate('abc', stringSchema)
// 'abc' (returns as is)

validate(5, stringSchema)
// '5' (converts to string)

validate({}, stringSchema)
// throws JoiValidationError
```

## convert

Does Joi-conversion, regardless of error/validity of value. Invalid values return `undefined`.

```ts
convert(5, stringSchema)
// '5' (valid, converted)

convert({}, stringSchema)
// undefined (invalid, does not throw)
```

## getValidationResult

- Validates with Joi.
- Returns `JoiValidationResult` with converted value and error (if any).
- Does not throw.

If `schema` is undefined - returns value as is.

```ts
getValidationResult('5', stringSchema)
// { value: '5', error: undefined }

getValidationResult({}, stringSchema)
// { error: ..., value: undefined }
```

## isValid

- Validates the value and returns `boolean`.
- Does not throw.
- Converted valid values are still considered valid.

```ts
isValid('5', stringSchema) // true
isValid(5, stringSchema) // still true, cause it can be converted to '5'
```

## undefinedIfInvalid

- Validates the value.
- Valid value is returned as is.
- Invalid value is returned as `undefined` without throwing an error.

```ts
undefinedIfInvalid('5', stringSchema)
// '5' as is, because it's valid

undefinedIfInvalid({}, stringSchema)
// undefined, because value is invalid
```

## JoiValidationResult

```ts
interface JoiValidationResult<T = any> {
  value: T
  error?: JoiValidationError
}
```

TODO: can be replaced by a Tuple of `[error, value]` in the future.

## JoiValidationError

```ts
import { ValidationErrorItem } from 'joi'

interface JoiValidationErrorData extends ErrorData {
  joiValidationErrorItems: ValidationErrorItem[]
  joiValidationObjectName?: string
  joiValidationObjectId?: string
}
```

## dateString

```ts
interface SomeObject {
  date: string // e.g `2021-06-21`
}

const schema = objectSchema<SomeObject>({
  date: stringSchema.dateString(),
  // same as:
  // date: dateStringSchema,
})
```

Options:

- `min`
- `max`

## dividable

Book example - number schema extension to test divide-ability.

```ts
const schema = numberSchema.dividable(15)
```

## Common validation schemas

Most of these are just conveniently defined/exported schemas, so you don't have to import `Joi`
object.

They are also conveniently typed, so validation functions can automatically defer the output type
based on these schemas (unlike default Joi typings that don't support "typed schemas" at all ATM).

- `booleanSchema`
- `stringSchema`
- `numberSchema`
- `integerSchema`
- `percentageSchema` - integer from 0 to 100 included
- `dateStringSchema`
- `binarySchema`
- `urlSchema` - defaults to require `https` protocol
- `arraySchema` - e.g: `arraySchema(stringSchema)`
- `objectSchema` - e.g `objectSchema<SomeInterface({ date: dateSchema })`
- `anySchema`
- `anyObjectSchema` - does `stripUnknown: false` hence permits any keys
- `idSchema` - 6-64 chars of alphanumeric plus underscore: `/^[a-z0-9_]*$/`
- `unixTimestampSchema` - allows values from 0 to a timestamp that represents 2500-01-01
- `emailSchema` - standard Joi's, does TLD validation by default, adds `lowercase()`
- `semVerSchema`
- `utcOffsetSchema`
- `ipAddressSchema`
