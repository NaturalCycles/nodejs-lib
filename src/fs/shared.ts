import * as fs from 'fs-extra'
import { boldWhite } from '../colors'

export interface GenGlobOptions {
  /**
   * @default to `**` (everything, including sub-dirs)
   */
  inputPatterns?: string[]

  /**
   * @default . (cwd)
   */
  baseDir: string

  /**
   * @default . (cwd)
   */
  outputDir: string
}

export function globPrepare(opt: GenGlobOptions): void {
  // Default pattern
  if (!opt.inputPatterns?.length) opt.inputPatterns = ['**']

  // default to cwd
  opt.baseDir ||= '.'
  opt.outputDir ||= '.'

  if (!fs.existsSync(opt.baseDir)) {
    console.log(`kpy: baseDir doesn't exist: ${boldWhite(opt.baseDir)}`)
    return
  }

  fs.ensureDirSync(opt.outputDir)
}
