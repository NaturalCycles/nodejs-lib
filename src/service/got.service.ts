import { GotJSONOptions } from 'got'
import * as got from 'got'
import * as http from 'http'
import { getDebug } from '../util/debug'
const debug = getDebug(__filename)

export interface GotOptions extends GotJSONOptions {}

export interface GotResponse<T = string> extends http.IncomingMessage {
  body: T
  url: string
  requestUrl: string
  fromCache: boolean
  redirectUrls?: string[]
}

class GotService {
  /**
   * Just a proxy method, to simplify importing `* as got`, etc.
   */
  async got<T = string> (url: string, opt: got.GotOptions<any> = {}): Promise<GotResponse<T>> {
    return got(url, opt) as any
  }

  async request<T = any> (method: string, url: string, _opt: Partial<GotOptions> = {}): Promise<T> {
    debug(`>> ${method} ${url}`)
    const opt: GotOptions = {
      method,
      json: true,
      ..._opt,
    }

    try {
      const r = await got(url, opt)
      debug(`<< ${method} ${url}`)
      return r.body
    } catch (err) {
      return this.handleError(err, method, url)
    }
  }

  private handleError (err: any, method: string, url: string): never {
    if (!err || !err.response) throw err

    const r = err.response
    const errBody = JSON.stringify(r.body || '<empty_body>', null, 2)
    const msg = `<< ERROR ${r.statusCode} ${method} ${url}\n${errBody}`
    const e: any = new Error(msg)
    e.response = r
    throw e
  }

  // convenience methods
  async get<T = any> (url: string, opt?: Partial<GotOptions>): Promise<T> {
    return this.request<T>('GET', url, opt)
  }
  async post<T = any> (url: string, opt?: Partial<GotOptions>): Promise<T> {
    return this.request<T>('POST', url, opt)
  }
  async put<T = any> (url: string, opt?: Partial<GotOptions>): Promise<T> {
    return this.request<T>('PUT', url, opt)
  }
  async delete<T = any> (url: string, opt?: Partial<GotOptions>): Promise<T> {
    return this.request<T>('DELETE', url, opt)
  }
}

export const gotService = new GotService()
