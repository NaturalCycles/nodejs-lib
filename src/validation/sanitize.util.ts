import * as sanitize from 'sanitize-html'

export type SanitizeHTMLOptions = sanitize.IOptions

/**
 * Simply a wrapper around `sanitize-html` library.
 *
 * @experimental
 */
export function sanitizeHTML(s: string, opt?: SanitizeHTMLOptions): string {
  return sanitize(s, opt)
}
