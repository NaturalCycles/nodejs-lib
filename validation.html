<!DOCTYPE html>
<html lang="en-US">
  <head>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width,initial-scale=1">
    <title>Validation | nodejs-lib</title>
    <meta name="generator" content="VuePress 1.9.7">
    
    <meta name="description" content="">
    
    <link rel="preload" href="/nodejs-lib/assets/css/0.styles.2bbbbdf1.css" as="style"><link rel="preload" href="/nodejs-lib/assets/js/app.b32b6a46.js" as="script"><link rel="preload" href="/nodejs-lib/assets/js/2.23454914.js" as="script"><link rel="preload" href="/nodejs-lib/assets/js/19.6b3f8047.js" as="script"><link rel="prefetch" href="/nodejs-lib/assets/js/10.b311f070.js"><link rel="prefetch" href="/nodejs-lib/assets/js/11.f6fae681.js"><link rel="prefetch" href="/nodejs-lib/assets/js/12.c8e173fe.js"><link rel="prefetch" href="/nodejs-lib/assets/js/13.fc07e4fa.js"><link rel="prefetch" href="/nodejs-lib/assets/js/14.d4814555.js"><link rel="prefetch" href="/nodejs-lib/assets/js/15.f2620e00.js"><link rel="prefetch" href="/nodejs-lib/assets/js/16.5b91f6c9.js"><link rel="prefetch" href="/nodejs-lib/assets/js/17.46546f60.js"><link rel="prefetch" href="/nodejs-lib/assets/js/18.13a7664e.js"><link rel="prefetch" href="/nodejs-lib/assets/js/3.4a6755ce.js"><link rel="prefetch" href="/nodejs-lib/assets/js/4.03071ad0.js"><link rel="prefetch" href="/nodejs-lib/assets/js/5.d2f70f50.js"><link rel="prefetch" href="/nodejs-lib/assets/js/6.2399b5ae.js"><link rel="prefetch" href="/nodejs-lib/assets/js/7.95714517.js"><link rel="prefetch" href="/nodejs-lib/assets/js/8.a8b22878.js"><link rel="prefetch" href="/nodejs-lib/assets/js/9.1561aaaf.js">
    <link rel="stylesheet" href="/nodejs-lib/assets/css/0.styles.2bbbbdf1.css">
  </head>
  <body>
    <div id="app" data-server-rendered="true"><div class="theme-container"><header class="navbar"><div class="sidebar-button"><svg xmlns="http://www.w3.org/2000/svg" aria-hidden="true" role="img" viewBox="0 0 448 512" class="icon"><path fill="currentColor" d="M436 124H12c-6.627 0-12-5.373-12-12V80c0-6.627 5.373-12 12-12h424c6.627 0 12 5.373 12 12v32c0 6.627-5.373 12-12 12zm0 160H12c-6.627 0-12-5.373-12-12v-32c0-6.627 5.373-12 12-12h424c6.627 0 12 5.373 12 12v32c0 6.627-5.373 12-12 12zm0 160H12c-6.627 0-12-5.373-12-12v-32c0-6.627 5.373-12 12-12h424c6.627 0 12 5.373 12 12v32c0 6.627-5.373 12-12 12z"></path></svg></div> <a href="/nodejs-lib/" class="home-link router-link-active"><!----> <span class="site-name">nodejs-lib</span></a> <div class="links"><div class="search-box"><input aria-label="Search" autocomplete="off" spellcheck="false" value=""> <!----></div> <nav class="nav-links can-hide"> <a href="https://github.com/NaturalCycles/nodejs-lib" target="_blank" rel="noopener noreferrer" class="repo-link">
    GitHub
    <span><svg xmlns="http://www.w3.org/2000/svg" aria-hidden="true" focusable="false" x="0px" y="0px" viewBox="0 0 100 100" width="15" height="15" class="icon outbound"><path fill="currentColor" d="M18.8,85.1h56l0,0c2.2,0,4-1.8,4-4v-32h-8v28h-48v-48h28v-8h-32l0,0c-2.2,0-4,1.8-4,4v56C14.8,83.3,16.6,85.1,18.8,85.1z"></path> <polygon fill="currentColor" points="45.7,48.7 51.3,54.3 77.2,28.5 77.2,37.2 85.2,37.2 85.2,14.9 62.8,14.9 62.8,22.9 71.5,22.9"></polygon></svg> <span class="sr-only">(opens new window)</span></span></a></nav></div></header> <div class="sidebar-mask"></div> <aside class="sidebar"><nav class="nav-links"> <a href="https://github.com/NaturalCycles/nodejs-lib" target="_blank" rel="noopener noreferrer" class="repo-link">
    GitHub
    <span><svg xmlns="http://www.w3.org/2000/svg" aria-hidden="true" focusable="false" x="0px" y="0px" viewBox="0 0 100 100" width="15" height="15" class="icon outbound"><path fill="currentColor" d="M18.8,85.1h56l0,0c2.2,0,4-1.8,4-4v-32h-8v28h-48v-48h28v-8h-32l0,0c-2.2,0-4,1.8-4,4v56C14.8,83.3,16.6,85.1,18.8,85.1z"></path> <polygon fill="currentColor" points="45.7,48.7 51.3,54.3 77.2,28.5 77.2,37.2 85.2,37.2 85.2,14.9 62.8,14.9 62.8,22.9 71.5,22.9"></polygon></svg> <span class="sr-only">(opens new window)</span></span></a></nav>  <ul class="sidebar-links"><li><section class="sidebar-group depth-0"><p class="sidebar-heading open"><span></span> <!----></p> <ul class="sidebar-links sidebar-group-items"><li><a href="/nodejs-lib/" aria-current="page" class="sidebar-link">nodejs-lib</a></li><li><a href="/nodejs-lib/validation.html" aria-current="page" class="active sidebar-link">Validation</a><ul class="sidebar-sub-headers"><li class="sidebar-sub-header"><a href="/nodejs-lib/validation.html#validate" class="sidebar-link">validate</a></li><li class="sidebar-sub-header"><a href="/nodejs-lib/validation.html#convert" class="sidebar-link">convert</a></li><li class="sidebar-sub-header"><a href="/nodejs-lib/validation.html#getvalidationresult" class="sidebar-link">getValidationResult</a></li><li class="sidebar-sub-header"><a href="/nodejs-lib/validation.html#isvalid" class="sidebar-link">isValid</a></li><li class="sidebar-sub-header"><a href="/nodejs-lib/validation.html#undefinedifinvalid" class="sidebar-link">undefinedIfInvalid</a></li><li class="sidebar-sub-header"><a href="/nodejs-lib/validation.html#joivalidationresult" class="sidebar-link">JoiValidationResult</a></li><li class="sidebar-sub-header"><a href="/nodejs-lib/validation.html#joivalidationerror" class="sidebar-link">JoiValidationError</a></li><li class="sidebar-sub-header"><a href="/nodejs-lib/validation.html#datestring" class="sidebar-link">dateString</a></li><li class="sidebar-sub-header"><a href="/nodejs-lib/validation.html#dividable" class="sidebar-link">dividable</a></li><li class="sidebar-sub-header"><a href="/nodejs-lib/validation.html#common-validation-schemas" class="sidebar-link">Common validation schemas</a></li></ul></li><li><a href="/nodejs-lib/http.html" class="sidebar-link">http</a></li><li><a href="/nodejs-lib/fs.html" class="sidebar-link">fs</a></li><li><a href="/nodejs-lib/stream.html" class="sidebar-link">Stream</a></li><li><a href="/nodejs-lib/colors.html" class="sidebar-link">Colors</a></li><li><a href="/nodejs-lib/debug.html" class="sidebar-link">Debug</a></li><li><a href="/nodejs-lib/exec.html" class="sidebar-link">Exec</a></li><li><a href="/nodejs-lib/secrets.html" class="sidebar-link">Secret management</a></li><li><a href="/nodejs-lib/security.html" class="sidebar-link">Security</a></li><li><a href="/nodejs-lib/script.html" class="sidebar-link">Script</a></li><li><a href="/nodejs-lib/slack.html" class="sidebar-link">Slack</a></li><li><a href="/nodejs-lib/other.html" class="sidebar-link">Other</a></li></ul></section></li></ul> </aside> <main class="page"> <div class="theme-default-content content__default"><h1 id="validation"><a href="#validation" class="header-anchor">#</a> Validation</h1> <p>Based on popular <a href="https://github.com/sideway/joi" target="_blank" rel="noopener noreferrer">Joi<span><svg xmlns="http://www.w3.org/2000/svg" aria-hidden="true" focusable="false" x="0px" y="0px" viewBox="0 0 100 100" width="15" height="15" class="icon outbound"><path fill="currentColor" d="M18.8,85.1h56l0,0c2.2,0,4-1.8,4-4v-32h-8v28h-48v-48h28v-8h-32l0,0c-2.2,0-4,1.8-4,4v56C14.8,83.3,16.6,85.1,18.8,85.1z"></path> <polygon fill="currentColor" points="45.7,48.7 51.3,54.3 77.2,28.5 77.2,37.2 85.2,37.2 85.2,14.9 62.8,14.9 62.8,22.9 71.5,22.9"></polygon></svg> <span class="sr-only">(opens new window)</span></span></a> validation library.</p> <p>Assumes that you don't need to have <code>joi</code> as your dependency, but just use convenience methods from
<code>nodejs-lib</code> (and use exported <code>Joi</code>, which is extended with all the extensions and defaults
documented below). Do not import <code>joi</code> directly, otherwise you lose these defaults and extensions.</p> <p>Default options:</p> <ul><li><code>presense: required</code> - all declared fields are required by default</li> <li><code>convert: true</code></li> <li><code>allowUnknown: true</code></li> <li><code>stripUnknown:</code> <ul><li><code>objects: true</code></li> <li><code>arrays: false</code></li></ul></li></ul> <p>Extensions:</p> <ul><li>Strings are trimmed by default</li> <li>Treats empty strings <code>'''</code> as empty/undefined</li> <li><code>dateString</code> stringSchema extension to validate ISO8601 dates</li> <li><code>dividable</code> numberSchema extension to validate divide-ability</li></ul> <p>Does some processing of the Error thrown:</p> <ul><li>Includes <code>objectName</code> (if provided), otherwise tries to detect &quot;classname&quot; (doing
<code>value?.constructor?.name</code>).</li> <li>Includes <code>objectId</code> if it can be detected (<code>id</code> field of an object passed). Useful e.g in <code>db-lib</code>
CommonDao methods.</li> <li>Truncates error message to 1000 characters.</li> <li>Truncates <code>err.details</code> to max 5 items</li></ul> <h2 id="validate"><a href="#validate" class="header-anchor">#</a> validate</h2> <ul><li>Validates with Joi.</li> <li>Throws JoiValidationError if invalid.</li> <li>Returns <em>converted</em> value.</li> <li>If <code>schema</code> is undefined - returns value as is.</li></ul> <div class="language-ts extra-class"><pre class="language-ts"><code><span class="token function">validate</span><span class="token punctuation">(</span><span class="token string">'abc'</span><span class="token punctuation">,</span> stringSchema<span class="token punctuation">)</span>
<span class="token comment">// 'abc' (returns as is)</span>

<span class="token function">validate</span><span class="token punctuation">(</span><span class="token number">5</span><span class="token punctuation">,</span> stringSchema<span class="token punctuation">)</span>
<span class="token comment">// '5' (converts to string)</span>

<span class="token function">validate</span><span class="token punctuation">(</span><span class="token punctuation">{</span><span class="token punctuation">}</span><span class="token punctuation">,</span> stringSchema<span class="token punctuation">)</span>
<span class="token comment">// throws JoiValidationError</span>
</code></pre></div><h2 id="convert"><a href="#convert" class="header-anchor">#</a> convert</h2> <p>Does Joi-conversion, regardless of error/validity of value. Invalid values return <code>undefined</code>.</p> <div class="language-ts extra-class"><pre class="language-ts"><code><span class="token function">convert</span><span class="token punctuation">(</span><span class="token number">5</span><span class="token punctuation">,</span> stringSchema<span class="token punctuation">)</span>
<span class="token comment">// '5' (valid, converted)</span>

<span class="token function">convert</span><span class="token punctuation">(</span><span class="token punctuation">{</span><span class="token punctuation">}</span><span class="token punctuation">,</span> stringSchema<span class="token punctuation">)</span>
<span class="token comment">// undefined (invalid, does not throw)</span>
</code></pre></div><h2 id="getvalidationresult"><a href="#getvalidationresult" class="header-anchor">#</a> getValidationResult</h2> <ul><li>Validates with Joi.</li> <li>Returns <code>JoiValidationResult</code> with converted value and error (if any).</li> <li>Does not throw.</li></ul> <p>If <code>schema</code> is undefined - returns value as is.</p> <div class="language-ts extra-class"><pre class="language-ts"><code><span class="token function">getValidationResult</span><span class="token punctuation">(</span><span class="token string">'5'</span><span class="token punctuation">,</span> stringSchema<span class="token punctuation">)</span>
<span class="token comment">// { value: '5', error: undefined }</span>

<span class="token function">getValidationResult</span><span class="token punctuation">(</span><span class="token punctuation">{</span><span class="token punctuation">}</span><span class="token punctuation">,</span> stringSchema<span class="token punctuation">)</span>
<span class="token comment">// { error: ..., value: undefined }</span>
</code></pre></div><h2 id="isvalid"><a href="#isvalid" class="header-anchor">#</a> isValid</h2> <ul><li>Validates the value and returns <code>boolean</code>.</li> <li>Does not throw.</li> <li>Converted valid values are still considered valid.</li></ul> <div class="language-ts extra-class"><pre class="language-ts"><code><span class="token function">isValid</span><span class="token punctuation">(</span><span class="token string">'5'</span><span class="token punctuation">,</span> stringSchema<span class="token punctuation">)</span> <span class="token comment">// true</span>
<span class="token function">isValid</span><span class="token punctuation">(</span><span class="token number">5</span><span class="token punctuation">,</span> stringSchema<span class="token punctuation">)</span> <span class="token comment">// still true, cause it can be converted to '5'</span>
</code></pre></div><h2 id="undefinedifinvalid"><a href="#undefinedifinvalid" class="header-anchor">#</a> undefinedIfInvalid</h2> <ul><li>Validates the value.</li> <li>Valid value is returned as is.</li> <li>Invalid value is returned as <code>undefined</code> without throwing an error.</li></ul> <div class="language-ts extra-class"><pre class="language-ts"><code><span class="token function">undefinedIfInvalid</span><span class="token punctuation">(</span><span class="token string">'5'</span><span class="token punctuation">,</span> stringSchema<span class="token punctuation">)</span>
<span class="token comment">// '5' as is, because it's valid</span>

<span class="token function">undefinedIfInvalid</span><span class="token punctuation">(</span><span class="token punctuation">{</span><span class="token punctuation">}</span><span class="token punctuation">,</span> stringSchema<span class="token punctuation">)</span>
<span class="token comment">// undefined, because value is invalid</span>
</code></pre></div><h2 id="joivalidationresult"><a href="#joivalidationresult" class="header-anchor">#</a> JoiValidationResult</h2> <div class="language-ts extra-class"><pre class="language-ts"><code><span class="token keyword">interface</span> <span class="token class-name">JoiValidationResult<span class="token operator">&lt;</span><span class="token constant">T</span> <span class="token operator">=</span> <span class="token builtin">any</span><span class="token operator">&gt;</span></span> <span class="token punctuation">{</span>
  value<span class="token operator">:</span> <span class="token constant">T</span>
  error<span class="token operator">?</span><span class="token operator">:</span> JoiValidationError
<span class="token punctuation">}</span>
</code></pre></div><p>TODO: can be replaced by a Tuple of <code>[error, value]</code> in the future.</p> <h2 id="joivalidationerror"><a href="#joivalidationerror" class="header-anchor">#</a> JoiValidationError</h2> <div class="language-ts extra-class"><pre class="language-ts"><code><span class="token keyword">import</span> <span class="token punctuation">{</span> ValidationErrorItem <span class="token punctuation">}</span> <span class="token keyword">from</span> <span class="token string">'joi'</span>

<span class="token keyword">interface</span> <span class="token class-name">JoiValidationErrorData</span> <span class="token keyword">extends</span> <span class="token class-name">ErrorData</span> <span class="token punctuation">{</span>
  joiValidationErrorItems<span class="token operator">:</span> ValidationErrorItem<span class="token punctuation">[</span><span class="token punctuation">]</span>
  joiValidationObjectName<span class="token operator">?</span><span class="token operator">:</span> <span class="token builtin">string</span>
  joiValidationObjectId<span class="token operator">?</span><span class="token operator">:</span> <span class="token builtin">string</span>
<span class="token punctuation">}</span>
</code></pre></div><h2 id="datestring"><a href="#datestring" class="header-anchor">#</a> dateString</h2> <div class="language-ts extra-class"><pre class="language-ts"><code><span class="token keyword">interface</span> <span class="token class-name">SomeObject</span> <span class="token punctuation">{</span>
  date<span class="token operator">:</span> <span class="token builtin">string</span> <span class="token comment">// e.g `2021-06-21`</span>
<span class="token punctuation">}</span>

<span class="token keyword">const</span> schema <span class="token operator">=</span> <span class="token generic-function"><span class="token function">objectSchema</span><span class="token generic class-name"><span class="token operator">&lt;</span>SomeObject<span class="token operator">&gt;</span></span></span><span class="token punctuation">(</span><span class="token punctuation">{</span>
  date<span class="token operator">:</span> stringSchema<span class="token punctuation">.</span><span class="token function">dateString</span><span class="token punctuation">(</span><span class="token punctuation">)</span><span class="token punctuation">,</span>
  <span class="token comment">// same as:</span>
  <span class="token comment">// date: dateStringSchema,</span>
<span class="token punctuation">}</span><span class="token punctuation">)</span>
</code></pre></div><p>Options:</p> <ul><li><code>min</code></li> <li><code>max</code></li></ul> <h2 id="dividable"><a href="#dividable" class="header-anchor">#</a> dividable</h2> <p>Book example - number schema extension to test divide-ability.</p> <div class="language-ts extra-class"><pre class="language-ts"><code><span class="token keyword">const</span> schema <span class="token operator">=</span> numberSchema<span class="token punctuation">.</span><span class="token function">dividable</span><span class="token punctuation">(</span><span class="token number">15</span><span class="token punctuation">)</span>
</code></pre></div><h2 id="common-validation-schemas"><a href="#common-validation-schemas" class="header-anchor">#</a> Common validation schemas</h2> <p>Most of these are just conveniently defined/exported schemas, so you don't have to import <code>Joi</code>
object.</p> <p>They are also conveniently typed, so validation functions can automatically defer the output type
based on these schemas (unlike default Joi typings that don't support &quot;typed schemas&quot; at all ATM).</p> <ul><li><code>booleanSchema</code></li> <li><code>stringSchema</code></li> <li><code>numberSchema</code></li> <li><code>integerSchema</code></li> <li><code>percentageSchema</code> - integer from 0 to 100 included</li> <li><code>dateStringSchema</code></li> <li><code>binarySchema</code></li> <li><code>urlSchema</code> - defaults to require <code>https</code> protocol</li> <li><code>arraySchema</code> - e.g: <code>arraySchema(stringSchema)</code></li> <li><code>objectSchema</code> - e.g <code>objectSchema&lt;SomeInterface({ date: dateSchema })</code></li> <li><code>anySchema</code></li> <li><code>anyObjectSchema</code> - does <code>stripUnknown: false</code> hence permits any keys</li> <li><code>idSchema</code> - 6-64 chars of alphanumeric plus underscore: <code>/^[a-z0-9_]*$/</code></li> <li><code>unixTimestampSchema</code> - allows values from 0 to a timestamp that represents 2500-01-01</li> <li><code>emailSchema</code> - standard Joi's, does TLD validation by default, adds <code>lowercase()</code></li> <li><code>semVerSchema</code></li> <li><code>utcOffsetSchema</code></li> <li><code>ipAddressSchema</code></li></ul></div> <footer class="page-edit"><!----> <!----></footer> <div class="page-nav"><p class="inner"><span class="prev">
      ←
      <a href="/nodejs-lib/" class="prev router-link-active">
        nodejs-lib
      </a></span> <span class="next"><a href="/nodejs-lib/http.html">
        http
      </a>
      →
    </span></p></div> </main></div><div class="global-ui"></div></div>
    <script src="/nodejs-lib/assets/js/app.b32b6a46.js" defer></script><script src="/nodejs-lib/assets/js/2.23454914.js" defer></script><script src="/nodejs-lib/assets/js/19.6b3f8047.js" defer></script>
  </body>
</html>
