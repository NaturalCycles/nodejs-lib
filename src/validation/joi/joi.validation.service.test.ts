import { stringSchema } from './joi.shared.schemas'
import { JoiValidationError } from './joi.validation.error'
import { joiValidationService } from './joi.validation.service'

const schema = {
  a1: stringSchema.min(2).max(5),
  a2: stringSchema.min(2).optional(),
}

const invalidValues: any[] = [
  undefined,
  null,
  '',
  'abc',
  5,
  () => 'a',
  { a1: 'a' },
  { a1: 'a12345' },
]

const validValues: any[] = [
  {
    a1: 'ff',
  },
  {
    a1: 'ff',
    a2: 'sdfer',
  },
]

test('should fail on invalid values', () => {
  invalidValues.forEach(v => {
    expect(() => joiValidationService.validate(v, schema)).toThrowErrorMatchingSnapshot()
  })
})

test('should pass on valid values', () => {
  validValues.forEach(v => {
    joiValidationService.validate(v, schema)
  })
})

test('should trim strings by default', async () => {
  const v = { a1: ' sdf ' }
  const v2 = joiValidationService.validate(v, schema)

  expect(v2.a1).toBe('sdf')
  expect(v === v2).toBeFalsy() // object should be cloned
})

test('should strip unknown keys', async () => {
  const v = {
    a1: 'ff',
    unk: 'ddd',
  }
  const v2 = joiValidationService.validate(v, schema)

  expect(v2).toEqual({ a1: 'ff' })
  expect(v2.unk).toBeUndefined()
})

test('getValidationResult should still convert', async () => {
  const v = {
    a1: ' ff ', // to be converted
    a2: 'a', // invalid!
  }
  const vr = joiValidationService.getValidationResult(v, schema, 'objName')
  expect(vr.value.a1).toBe('ff')
  expect(vr.error).toBeInstanceOf(JoiValidationError)
  expect(vr.error).toMatchSnapshot()
})

test('getValidationResult valid', async () => {
  const vr = joiValidationService.getValidationResult('asd', stringSchema)
  expect(vr.error).toBeUndefined()
})

test('validationErrorToString', async () => {
  expect(joiValidationService.validationErrorToString(undefined as any)).toBeUndefined()
})

test('error should contain errorItems', async () => {
  const v = {
    a1: ' ff ', // to be converted
    a2: 'a', // invalid!
  }
  const { error } = joiValidationService.getValidationResult(v, schema, 'objName')
  expect(error!.data).toMatchSnapshot()
})

// optional
// default values
