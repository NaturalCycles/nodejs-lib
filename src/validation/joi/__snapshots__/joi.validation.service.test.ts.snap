// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getValidationResult should still convert 1`] = `
[AppValidationError: objName
{
  "a1": " ff ",
  "a2" [31m[1][0m: "a"
}
[31m
[1] "a2" length must be at least 2 characters long[0m]
`;
